ramls/tenant.raml

```diff
@@ -2,68 +2,79 @@
 
 title: Tenant API
 baseUri: http://localhost:8081/{version}
-version: v1
+version: v2
 
 traits:
   validate: !include ../traits/validation.raml
+  internal: !include ../traits/internal-error.raml
 types:
-  tenantAttributes: !include ../schemas/moduleInfo.schema
+  tenantAttributes: !include ../schemas/tenantAttributes.schema
+  tenantJob: !include ../schemas/tenantJob.schema
   errors: !include ../schemas/errors.schema
 
 /_/tenant:
   post:
-    description: Provision a new / update an existing tenant
-    is: [validate]
+    description: Create tenant job (create, upgrade, delete)
+    is: [validate,internal]
     body:
       application/json:
         type: tenantAttributes
         example:
-          strict: false
-          value: !include ../examples/moduleInfo.sample
+          value: !include ../examples/tenantAttributes.sample
     responses:
-      200:
-        description: "tenant upgraded successfully to new module ID"
-        body:
-          application/json:
       201:
-        description: "tenant enabled successfully for the module ID"
+        description: "Tenant job created"
         body:
           application/json:
-      204:
-        description: "tenant enabled successfully for the module ID (there was nothing to do)"
+            type: tenantJob
+            example:
+              value: !include ../examples/tenantJob.sample
+        headers:
+          Location:
+            description: URI of the job
       400:
-        description: "Bad request or failing operation; SQL commands that failed when initalizing or upgrading the tenant are listed as JSON array elements."
-        body:
-          text/plain:
-      500:
-        description: "Internal server error"
-        body:
-          text/plain:
-            example: "Internal server error, contact administrator"
-  get:
-    description: Does tenant id already exist
-    responses:
-      200:
-        description: "true or false indicator"
-        body:
-          text/plain:
-            example: "Text....."
-      500:
-        description: "Internal server error"
-        body:
-          text/plain:
-            example: "Internal server error, contact administrator"
-  delete:
```

Comment:
Since RMB 30.0.0 we store the module version in the database:
https://github.com/folio-org/raml-module-builder/blob/v30.0.0/domain-models-runtime/src/main/resources/templates/db_scripts/main.ftl#L210
Therefore there is no longer a need to require the module_from parameter.
We also don't need the module_to parameter because the module knowns its current version that we run (for creation or for upgrading to).

I suggest to keep both the POST and DELETE endpoint. They can be used without module_from and module_to parameters:
POST `/_/tenant` creates if there is no database schema yet.
POST `/_/tenant` upgrades if the the database schema already exists.
DELETE `/_/tenant` deletes the database schema.
It might be useful to have
GET `/_/tenant` that returns 404 if no database schema exists, otherwise returns "moduleId" and "dataModuleId". "dataModuleId" is optional and is only used for storage modules. If "moduleId" and "dataModuleId" differ then a POST `/_/tenant` is needed to upgrade the database.

Reply:
module_to and module_from may be ignored at the moment.. But they are still passed to SQL scripts (if they are not actually in use, that code should be removed). They were in use at a point, so it might be in used by another implementation than just RMB. So the motivation for removing them because RMB may not use them is not really an argument. In particular any module that doesn't use storage will not know module_from

RMB in fact at the moment has weird semantics when a module is disabled (module_from given, but no module_to). It does nothing.. It most certainly does not disable the tenant. It should at least disable the tenant by some means.. Of course not purge data. This is what happens on `apt remove` in Debian.

As for the DELETE and keeping that. I thought it would be pointless. Mostly because it's not async. RMB might be fast for purge, but in the generic case. So that's why PURGE is just another Tenant Job. And the implementation only has to implement one POST method.

Remember that the initial design of the tenant API was CRUD for the Tenant. Now it's CRUD for the Tenant job "operation". That's also why both GET and DELETE would not be part of a new CRUD job API to get tenant or remove it. It is to get job status and delete job.

As for old tenant API: the GET was never used by Okapi. Not to mention that it is lying. If anything it should safe false if a Tenant is disabled. It doesn't. It only returns NOT FOUND if tenant is purged earlier. Again.. The existing design was CRUD for Tenant.

So to summarize I don't see a good argument for changing the proposed async API. What is missing though is a more fined tuned Tenant Job API to be able to support something ala 3-way commit (later).

Reply:
If a module does not have any storage there is nothing it can migrate, and no need to know module_from.